### PR TITLE
Storage policies

### PR DESCRIPTION
**Pull request by carycheng**

### CODE REVIEW

**Comment by boxcla** [comment]
Hi @carycheng, thanks for the pull request. Before we can merge it, we need you to sign our Contributor License Agreement. You can do so electronically here: http://opensource.box.com/cla

Once you have signed, just add a comment to this pull request saying, "CLA signed". Thanks!

File: boxsdk/client/client.py [hunk 1]
```diff
@@ -13,8 +13,12 @@
 from ..object.file import File
 from ..object.group import Group
 from ..object.group_membership import GroupMembership
+from ..object.storage_policy import StoragePolicy
+from ..object.storage_policy_assignment import StoragePolicyAssignment
 from ..util.shared_link import get_shared_link_header
 from ..util.translator import Translator
+from ..pagination.marker_based_object_collection import MarkerBasedObjectCollection
+
```
Comment (anthonywee) [inline boxsdk/client/client.py]: white space
alphabetical imports

File: boxsdk/client/client.py [hunk 2]
```diff
@@ -391,3 +395,107 @@ def get_url(self, endpoint, *args):
         """
         # pylint:disable=no-self-use
         return self._session.get_url(endpoint, *args)
+
```
Comment (anthonywee) [inline boxsdk/client/client.py]: white space

File: boxsdk/client/client.py [hunk 3]
```diff
@@ -391,3 +395,107 @@ def get_url(self, endpoint, *args):
...
+
+    def storage_policy(self, policy_id):
+        """
+        Initialize a :class: `StoragePolicy` object, whose box id is policy_id.
```
Comment (anthonywee) [inline boxsdk/client/client.py]: extra space after :class:

File: boxsdk/client/client.py [hunk 4]
```diff
@@ -407,3 +410,102 @@ def get_url(self, endpoint, *args):
...
+        :return:
+            A :class:`StoragePolicyAssignment` object with the given entry ID.
+        :rtype:
+            :class:`StoragePolicyAssignment`
+        """
+        return StoragePolicyAssignment(session=self._session, object_id=assignment_id)
+
+    def storage_policies(self, limit=None, marker=None, fields=None):
+        """
+        Get the entries in the storage policy using limit-offset paging.
```
Comment (anthonywee) [inline boxsdk/client/client.py]: marker based paging?

File: boxsdk/client/client.py [hunk 5]
```diff
@@ -407,3 +410,102 @@ def get_url(self, endpoint, *args):
...
+            url=self.get_url('storage_policies'),
+            limit=limit,
+            marker=marker,
+            fields=fields,
+            return_full_pages=False,
+        )
+
+    def storage_policy_assignments(self, resolved_for_type, resolved_for_id, marker=None, fields=None):
+        """
+        Get the entries in the storage policy assignment using limit-offset paging.
```
Comment (anthonywee) [inline boxsdk/client/client.py]: marker based paging

File: boxsdk/client/client.py [hunk 6]
```diff
@@ -407,3 +410,102 @@ def get_url(self, endpoint, *args):
...
+
+        :param resolved_for_type:
+            Set to either `user` or `enterprise`
+        :type limit:
+            unicode
```
Comment (anthonywee) [inline boxsdk/client/client.py]: add the back-ticks "`"

File: boxsdk/client/client.py [hunk 7]
```diff
@@ -407,3 +410,102 @@ def get_url(self, endpoint, *args):
...
+        :param resolved_for_id:
+            The id of the user or enterprise
+        :type limit:
+            unicode
```
Comment (anthonywee) [inline boxsdk/client/client.py]: back ticks

File: boxsdk/pagination/page.py [hunk 8]
```diff
@@ -38,7 +39,7 @@ def _translator(self):
 
         :rtype:   :class:`Translator`
         """
-        return self._session.translator
+        return Translator()
```
Comment (anthonywee) [inline boxsdk/pagination/page.py]: Why is it okay/necessary to use a new Translator() here? Maybe a comment could help?

File: boxsdk/object/storage_policy_assignment.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+from __future__ import unicode_literals, absolute_import
+
+from .base_object import BaseObject
+
+
+class StoragePolicyAssignment(BaseObject):
+    """Represents the storage policy assignment"""
+
+    def get_url(self, *args):
```
Comment (anthonywee) [inline boxsdk/object/storage_policy_assignment.py]: doc string

File: boxsdk/object/storage_policy.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+class StoragePolicy(BaseObject):
+    """Represents the storage policy"""
+
+    def get_url(self, *args):
+        """
+        Get url for storage policies.
+        """
+        return self._session.get_url('storage_policies', self._object_id, *args)
+
+    _item_type = 'storage_policy'
```
Comment (anthonywee) [inline boxsdk/object/storage_policy.py]: Maybe nicer to have _item_type at the top of the class, similar to other object classes?

File: boxsdk/object/storage_policy.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+    def assign(self, item):
+        url = self._session.get_url('storage_policy_assignments')
+        body = {
+            'storage_policy': {
+                'type': 'storage_policy',
+                'id': self.object_id
```
Comment (anthonywee) [inline boxsdk/object/storage_policy.py]: trailing comma

File: boxsdk/object/storage_policy.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+            },
+            'assigned_to': {
+                'type': item.object_type,
+                'id': item.object_id
```
Comment (anthonywee) [inline boxsdk/object/storage_policy.py]: trailing comma

File: test/unit/client/test_client.py [hunk 13]
```diff
@@ -324,3 +326,49 @@ def test_create_enterprise_user_returns_the_correct_user_object(mock_client, moc
...
     assert new_user.object_id == 1234
     assert new_user.name == test_user_name
+
+
+def test_get_storage_policies(mock_client, mock_box_session):
+    expected_url = mock_box_session.get_url('storage_policies')
+    mock_policy = {
+        'type': 'storage_policy',
+        'id': '12345',
+        'name': 'Test Storage Policy'
```
Comment (anthonywee) [inline test/unit/client/test_client.py]: trailing comma

File: test/unit/client/test_client.py [hunk 14]
```diff
@@ -324,3 +326,49 @@ def test_create_enterprise_user_returns_the_correct_user_object(mock_client, moc
...
+    }
+    mock_box_session.get.return_value.json.return_value = {
+        'limit': 100,
+        'entries': [mock_policy]
```
Comment (anthonywee) [inline test/unit/client/test_client.py]: trailing comma

File: test/unit/client/test_client.py [hunk 15]
```diff
@@ -324,3 +326,49 @@ def test_create_enterprise_user_returns_the_correct_user_object(mock_client, moc
...
+    resolved_for_type = 'user'
+    resolved_for_id = '1234'
+    expected_url = mock_box_session.get_url('storage_policy_assignments')
+    mock_assignment = {
+        'type': 'storage_policy_assignment',
+        'id': '12345',
+    }
+    mock_box_session.get.return_value.json.return_value = {
+        'limit': 100,
+        'entries': [mock_assignment]
```
Comment (anthonywee) [inline test/unit/client/test_client.py]: trailing comma

File: test/unit/object/test_storage_policy.py [hunk 16]
```diff
@@ -0,0 +1,49 @@
+# coding: utf-8
+
+from __future__ import unicode_literals
```
Comment (anthonywee) [inline test/unit/object/test_storage_policy.py]: absolute_import

File: test/unit/object/test_storage_policy.py [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    mock_box_session.get.assert_called_once_with(expected_url, headers=None, params=None)
+    assert isinstance(storage_policy, StoragePolicy)
+
+
+def test_assign(test_storage_policy, mock_user, mock_box_session):
+    expected_url = mock_box_session.get_url('storage_policy_assignments')
+    expected_data = {
+        'storage_policy': {
+            'type': 'storage_policy',
+            'id': '42'
```
Comment (anthonywee) [inline test/unit/object/test_storage_policy.py]: trailing comma

File: test/unit/object/test_storage_policy.py [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+        },
+        'assigned_to': {
+            'type': 'user',
+            'id': 'fake-user-100'
```
Comment (anthonywee) [inline test/unit/object/test_storage_policy.py]: trailing comma

File: test/unit/object/test_storage_policy.py [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+        }
```
Comment (anthonywee) [inline test/unit/object/test_storage_policy.py]: trailing comma

File: test/unit/object/test_storage_policy.py [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    mock_assignment = {
+        'type': 'storage_policy_assignment',
+        'id': '1234',
+        'storage_policy': {
+            'type': 'storage_policy',
+            'id': '1111'
```
Comment (anthonywee) [inline test/unit/object/test_storage_policy.py]: trailing comma
also on line below

File: test/unit/object/test_storage_policy_assignment.py [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+    mock_box_session.put.assert_called_once_with(expected_url, data=json.dumps(data), headers=None, params=None)
+    assert isinstance(storage_policy_assignment, StoragePolicyAssignment)
+
+
+def test_delete(test_storage_policy_assignment, delete_assignment_response, mock_box_session):
+    # pylint:disable=redefined-outer-name
+    mock_box_session.delete.return_value = delete_assignment_response
+    response = test_storage_policy_assignment.delete()
+
+    # pylint:disable=protected-access
```
Comment (anthonywee) [inline test/unit/object/test_storage_policy_assignment.py]: I don't see the protected access?

File: test/unit/object/test_storage_policy_assignment.py [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+            'type': 'storage_policy',
+            'id': new_policy_id,
+        }
+    }
+    mock_box_session.put.assert_called_once_with(expected_url, data=json.dumps(data), headers=None, params=None)
+    assert isinstance(storage_policy_assignment, StoragePolicyAssignment)
+
+
+def test_delete(test_storage_policy_assignment, delete_assignment_response, mock_box_session):
+    # pylint:disable=redefined-outer-name
```
Comment (anthonywee) [inline test/unit/object/test_storage_policy_assignment.py]: which outer name is being redefined?

**Pull request merged**